### DISCUSSION TITLE
Log's Request Id was duplicated on different server

### DISCUSSION THREAD

**Discussion by dengyakui** [datalust/seq-tickets#1416] [Help] [answered]
Today, when I was fixing some bugs,   found that the log's RequestID I searched was duplicated on two servers. By the logging timestamp, It's two different API requests from the client, should not be supposed to be a concurrent problem.
So I want to know if it was's Serilog's problem,  and How to make sure my Log's RequestID is Globally unique.

![image](https://user-images.githubusercontent.com/3945919/146721797-545b865f-4566-4277-946d-74bac126c962.png)

**Comment by liammclennan** [comment]
Seq will have stored what it received, so the question is how did you end up with two different requests on two different servers sending the same RequestId? It should not be possible to randomly generate the same guid twice so it would be interesting to look into how the RequestId is being set. If you'd like to share the code you use to configure serilog-aspnetcore I can have a look.

**Reply by dengyakui** [reply]
@liammclennan  I will give you a zip file,  my project is running on aspnetcore6.0, and still used startup class.
By the way, I checked lasted 1000 logs' RequestId writeen by 2 servers(136/137), and do an intersect to check duplicate RequestId, but not found duplicate logId
[log-config.zip](https://github.com/datalust/seq-tickets/files/7751655/log-config.zip)

**Reply by liammclennan** [reply]
@dengyakui unfortunately I don't have a viewer for deic files.

**Reply by dengyakui** [reply]
Hi, @liammclennan , sorry for the HEIC file, it was exported from my iPad. I transformed all the images to PNG format.
[log-config-v2.zip](https://github.com/datalust/seq-tickets/files/7759309/log-config-v2.zip)

**Comment by nblumhardt** [comment, marked as answer]
This is a really interesting issue! I think the problem is that, although it looks like a GUID, `RequestId` is not always one.

The most recent thread I can find points to:

https://github.com/aspnet/KestrelHttpServer/blob/a48222378b8249a26b093b5b835001c7c7b45815/src/Kestrel.Core/Internal/Infrastructure/CorrelationIdGenerator.cs

and:

https://github.com/aspnet/KestrelHttpServer/pull/1784

which is Kestrel code, though there's a comment mentioning that _Hosting_ uses a similar algorithm.

The request id is only expected to be unique per application instance, at least in the configuration above. Unlikely as a collision seems to be, at least on the surface, I don't think there's necessarily an issue here.

Others have [noted this](https://stackoverflow.com/questions/61685592/how-unique-is-httpcontext-traceidentifier-in-asp-net-core) too:

> after sending a ton of requests, I noticed that the generated values are similar

However, you can set [`Request.TraceIdentifier`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.httpcontext.traceidentifier?view=aspnetcore-6.0) in middleware to anything you like, so you could use some middleware at the beginning of the pipeline to set the identifier to a real GUID and avoid future collisions.

BTW, `UseSerilog()` should be called directly on the `IHostBuilder`, and not inside `ConfigureWebHostDefaults()` (legacy API). Hope this helps!

**Reply by dengyakui** [reply]
Hi， @nblumhardt ，many thanks for your response. According to above info, I think i can use a custom middleware to populate the RequestID’s value when the ID conflicts more.
